src/compiler/java_compiler.nit

```diff
@@ -0,0 +1,2297 @@
+					self.add("{res} = {value1} == {value2}; /* t2 is primitive and t1 can be */")
+					return res
+				end
+				test.add("{value1}.rtclass == {value2}.rtclass")
+			else
+				incompatible = true
+			end
+		else
+			primitive = null
+		end
+
+		if incompatible then
+			if maybe_null then
+				self.add("{res} = {value1} == {value2}; /* incompatible types {t1} vs. {t2}; but may be NULL*/")
+				return res
+			else
+				self.add("{res} = false; /* incompatible types {t1} vs. {t2}; cannot be NULL */")
+				return res
+			end
+		end
+		if primitive != null then
+			if primitive.is_java_primitive then
+				self.add("{res} = {value1} == {value2};")
+				return res
+			end
+			test.add("({value1}.value == {value2}.value")
+		else if can_be_primitive(value1) and can_be_primitive(value2) then
+			test.add("{value1}.rtclass == {value2}.rtclass")
+			var s = new Array[String]
+			for b in compiler.box_kinds do
+				var rt_name = b.mclass.rt_name
+				s.add "({value1}.rtclass == {rt_name}.get{rt_name}()) && ({value1}.value.equals({value2}.value))"
+				if b.mclass.name == "Float" then
+					s.add "({value1}.rtclass == RTClass_kernel_Float.getRTClass_kernel_Float() && {value1}.rtclass == {value2}.rtclass && Math.abs((double)({value1}.value)) == 0.0 && Math.abs((double)({value2}.value)) == 0.0)"
+				end
+			end
+			if s.is_empty then
+				self.add("{res} = {value1} == {value2}; /* both can be primitive */")
+				return res
+			end
+			test.add("({s.join(" || ")})")
+		else
+			self.add("{res} = {value1} == {value2}; /* no primitives */")
+			return res
+		end
+		self.add("{res} = {value1} == {value2} || ({test.join(" && ")});")
+		return res
+	end
+
+	# Attributes
+
+	# Generate a polymorphic attribute is_set test
+	fun isset_attribute(a: MAttribute, recv: RuntimeVariable): RuntimeVariable do
+		# TODO self.check_recv_notnull(recv)
+		var res = new_var(compiler.mainmodule.bool_type)
+
+		# What is the declared type of the attribute?
+		var mtype = a.intro.static_mtype.as(not null)
+		var intromclassdef = a.intro.mclassdef
+		mtype = mtype.resolve_for(intromclassdef.bound_mtype, intromclassdef.bound_mtype, intromclassdef.mmodule, true)
+
+		if mtype isa MNullableType then
+			add("{res} = true; /* easy isset: {a} on {recv.inspect} */")
+			return res
+		end
+		add("{res} = {recv}.attrs.get(\"{a.jname}\") != null; /* {a} on {recv.inspect} */")
+		return res
+	end
+
+	# Generate a polymorphic attribute read
+	fun read_attribute(a: MAttribute, recv: RuntimeVariable): RuntimeVariable do
+		# TODO check_recv_notnull(recv)
+		# TODO compile_check(v)
+		# What is the declared type of the attribute?
+		var ret = a.intro.static_mtype.as(not null)
+		var intromclassdef = a.intro.mclassdef
+		ret = ret.resolve_for(intromclassdef.bound_mtype, intromclassdef.bound_mtype, intromclassdef.mmodule, true)
+
+		# Check for Uninitialized attribute
+		if not ret isa MNullableType then check_attribute(a, recv)
+
+		return new_expr("{recv}.attrs.get(\"{a.jname}\")", ret)
```

Comment:
This line seems buggy: the string contains a Java RTVal but the you create the RuntimeVariable with `ret`. Therefore, if ret is a primitive type like Int, the RuntimeVariable will may be associated a java primitie tyle like int.
Either you should force the unboxing to extract the value, or consider that the attribute is returned as is (ie a boxed RuntimeVariable).

For the former solution, the following code should do the trick:

~~~nit
var object_type = compiler.mainmodule.object_type
var res = new_expr("{recv}.attrs.get(\"{a.jname}\")", object_type)
res = autobox(res, ret) # force unboxing of primitive type
return res
~~~

For the latter, use
~~~nit
var object_type = compiler.mainmodule.object_type
var res = new_expr("{recv}.attrs.get(\"{a.jname}\")", object_type)
res.mcasttype = ret # The Nit static type of the box is known
return res
~~~